[improve][ci] Add workflow for python local install

RobertIndie opened the pull request:
<!--
### Contribution Checklist
  
  - PR title format should be *[type][component] summary*. For details, see *[Guideline - Pulsar PR Naming Convention](https://docs.google.com/document/d/1d8Pw6ZbWk-_pCKdOmdvx9rnhPiyuxwq60_TrD68d7BA/edit#heading=h.trs9rsex3xom)*. 

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.

**(The sections below can be removed for hotfixes of typos)**
-->

### Motivation

Currently, when building the python client according to the tutorial: https://pulsar.apache.org/docs/next/client-libraries-python#install-from-source, it may not find some required dependencies like `fastavro` and `certifi`. 

### Modifications

* Use `pip install .` to install local python client
* Add workflow to verify local installation

### Documentation

Check the box below or label this PR directly.

Need to update docs? 

- [ ] `doc-required` 
(Your PR needs to update docs and you will update later)
  
- [x] `doc-not-needed` 
(Please explain why)
  
- [ ] `doc` 
(Your PR contains doc changes)

- [ ] `doc-complete`
(Docs have been already added)

tisonkun reviewed (commented):
https://github.com/apache/pulsar/blob/6dffa1d92f98e8f275abb6af7a6fe52ab425ae78/pulsar-client-cpp/python/setup.py#L75-L101

I don't know whether Python can resolve and install dependencies from `setup.py`, but it seems all required packages is beyond the current `requirements.txt`. For example, it can requires `grpcio`.

RobertIndie commented:
> I don't know whether Python can resolve and install dependencies from `setup.py`, but it seems all required packages is beyond the current `requirements.txt`. For example, it can requires `grpcio`.

Good question. I investigated it and found that we need to speficy the `setup_requires` to make it automatically install dependencies when installing the python binding. Like
```python
setup(
    name=NAME,
    version=VERSION,
    packages=['pulsar', 'pulsar.schema', 'pulsar.functions'],
    cmdclass={'build_ext': my_build_ext},
    ext_modules=[Extension('_pulsar', [])],

    author="Pulsar Devs",
    author_email="dev@pulsar.apache.org",
    description="Apache Pulsar Python client library",
    license="Apache License v2.0",
    url="https://pulsar.apache.org/",
    install_requires=dependencies,
    setup_requires=dependencies,   # <-  Need to specify setup_requires
    extras_require=extras_require,
)
```
But I don't know how to make it automatically install the required minimum dependencies when installing other modules. Like `avro` and `functions`. 
And the minimum dependency is actually only `certifi`.

Any ideas on this? /cc @BewareMyPower @shibd

BewareMyPower reviewed (changes requested):
I think we should modify the `setup.py` instead of adding a `requirements.txt`. We should add a workflow to build Python client from source (maybe in a docker container) and run a very simple python script to verify it.

RobertIndie commented:
Hi, I have changed this PR to use `pip install .` command to install the python client locally. And added a workflow to verify it. PTAL again. @tisonkun @BewareMyPower @shibd

tisonkun reviewed (approved):
LGTM. Thanks!

tisonkun commented:
cc @nodece could you help with reviewing this patch?

nodece reviewed (approved):
LGTM

tisonkun commented on .github/workflows/ci-python-build-3.9-client.yaml:
```diff
@@ -66,3 +66,13 @@ jobs:
         run: |
           echo "Build Python3.9 client library"
           pulsar-client-cpp/docker-build-python3.9.sh
+
+      - name: Local build python client
+        if: ${{ steps.check_changes.outputs.docs_only != 'true' }}
+        uses: actions/setup-python@v4
+        with:
+          python-version: '3.9'
+        working-directory: pulsar-client-cpp/python
+        run: |
+          echo "Local build python client library"
+          pip install .
```
This section is incorrect and produces:

```
a step cannot have both the `uses` and `run` keys
```

See https://github.com/apache/pulsar/actions/runs/2978090895.

I guess it will cause all PRs to fail. Please take a look @nodece @RobertIndie.

Weird why it's not failed on PR CI workflows.

tisonkun replied:
Filed an issue in https://github.com/apache/pulsar/issues/17430.

tisonkun replied:
fix https://github.com/apache/pulsar/pull/17431

nodece replied:
Good catch!

tisonkun replied:
I just wonder why it's not verified within this patch :)

nodece replied:
GitHub Actions didn't run this workflow.

tisonkun replied:
Yes. But why? The trigger section writes:

```yml
on:
  pull_request:
    branches:
      - master
    paths:
      - '.github/workflows/**'
      - 'pulsar-client-cpp/**'
```
